### ISSUE TITLE
[Support]: More functionalities

### ISSUE THREAD

**Issue by dsm150** [sinamics/ztnet#50]
### Describe the problem you are having

Hello
You have created a beautiful UI for the controller, in my opinion, it is more pleasant and more convenient than the original one.
But it lacks a few important features:
1. DNS setting for "Default router override" mode
2. Client settings: "Allow Ethernet Bridging" and "Do not auto-assign IPs"
3. And of course "Flow rules"
Is there any hope for their appearance in your project?

### Version

0.0.23

### Operating system

Debian

**Comment by sinamics** [comment]
Hi
Thank you for the feedback. Some of these are on my roadmap but i have not had the time to add it yet. 
But now as you have pointed these out i will try to speed up the progress. 
However, im not that familiar with Flow-Rules, but i guess its similar to firewall / iptables, will need to read up on that. 

You are also welcome to provide a PR if you like :)

**Comment by dsm150** [comment]
Yes, flow rules is great thing! Must have, if you manage netwoks with more then 5-10 hosts ))))
I will be very grateful to you for adding functions.
Ready to help you in any way I can. I have a fairly large network of devices)

**Comment by sinamics** [comment]
@dsm150 can you test version v0.1.1 and let me know what you think.

**Comment by dsm150** [comment]
with pleasure! But I'm not good at docker yet. To upgrade, docker compose pull && docker compose up -d is enough?

**Comment by dsm150** [comment]
the docker compose pull and docker compose up -d commands worked without errors, but the web interface stopped opening

**Comment by sinamics** [comment]
Those commands seems correct. 
what does `docker logs ztnet` provides?

**Comment by dsm150** [comment]
2 migrations found in prisma/migrations

Error: P3009

migrate found failed migrations in the target database, new migrations will not be applied. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve
The `20230714081023_flow` migration started at 2023-07-15 08:14:51.056683 UTC failed


Applying migrations to the database...
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "postgres:5432"

**Comment by dsm150** [comment]
I completely removed all containers and partitions, and reinstalled everything. Everything worked. I will test the added functionality. So far everything looks amazing!

**Comment by dsm150** [comment]
Tested:

- IP manual assign - **OK**
- Custom DNS server - **OK** but it's not intuitive to add the server's IP address - you need to press enter to add the server. An "add" button would be clearer. But that doesn't bother me at all.
- Client settings - **OK**
- Flow rules 
For example - I use this set of rules in order to isolate clients from each other, and so that only certain clients can see the rest:
```shell
# Allow only IPv4, IPv4 ARP, and IPv6 Ethernet frames. Typical default
drop
not ethertype ipv4
and not ethertype arp
and not ethertype ipv6
;

# Is this member a server?
tag server
id 2
enum 0 No
enum 1 Yes
default Yes
;

# if both members are not servers
break
not tor server 1
;

# Accept anything else. This is required since default is 'drop'.
accept;
``` 
This flow add "Tags" in client settings for each client, where i can select a role: Server - sees everyone, non-server - does not see anyone except the server
In our case, after add flow rule, can't see "Tags" in client settings.
![Screenshot_1](https://github.com/sinamics/next_ztnet/assets/56016218/3d3cb078-0204-4bd0-b3f2-2d5a046d468d)

**Comment by sinamics** [comment]
Thx for the testign and feedback.
I've attempted to replicate the database migration error you experienced, but so far, all updates are executing flawlessly on my end. I will conduct further testing, as the docker container update process should be seamless.

> Custom DNS server - OK but it's not intuitive to add the server's IP address - you need to press enter to add the server. An "add" button would be clearer. But that doesn't bother me at all.

I agree about adding a button for the dns would improve the UX.

> This flow add "Tags" in client settings for each client, where i can select a role: Server - sees everyone, non-server - does not see anyone except the server
> In our case, after add flow rule, can't see "Tags" in client settings.

i was not aware that the Capabilities and Tags sections had any user interactions.
Will add these as well..

**Comment by dsm150** [comment]
We are looking forward to the update.
You have done a titanic job! Already at this stage, your solution bypasses in terms of convenience and capabilities, the solution from dec0dOS (https://github.com/dec0dOS/zero-ui)

**Comment by sinamics** [comment]
@dsm150 Can you try v0.1.2. 
I think i got it right :)

This version has also db migration files that needs to be updated. I tried to update locally and i did not see any errors.
So just in case you experience any issues as last time, you might need to do a full reset. 

just do these commands:
`docker-compose pull`  and then `docker-compose up -d`

You also need to re-save your Flow-Rules to update the db with new values.

**Comment by dsm150** [comment]
@sinamics Amazing work!!!!
update went without errors, everything works properly!
Flow rules now work as expected.
I noticed one more thing - the member name field does not work correctly. If I enter the name of a peer, it may jump to a neighboring one or duplicate it to neighboring ones.

**Comment by sinamics** [comment]
> I noticed one more thing - the member name field does not work correctly. If I enter the name of a peer, it may jump to a neighboring one or duplicate it to neighboring ones.

yeah, i was only testing with one member. 
Will fix that.

**Comment by dsm150** [comment]
I took the liberty of publishing a post about your project in the official zerotier community. I think colleagues should know about such a wonderful project.

**Comment by sinamics** [comment]
thx. I hope it will be usable for someone. 
There is a lot of work remaining, but we will get there eventually

**Comment by dsm150** [comment]
If there is an opportunity to "buy you a cup of coffee" I will gladly do it. I'm from Russia, I can't make a direct transfer, but if you have an "electric" wallet (USDT,BTC,etc), I would really like to thank you.

**Comment by sinamics** [comment]
I appreciate your offer, but I currently do not have the option.
Thank you for asking.

**Comment by sinamics** [comment]
I made a few overvall improvements, and i think the peer name issue is resolved. 
v0.1.3

**Comment by dsm150** [comment]
The update was successful.
Looks like the problem with client names is gone.
The icing on the cake would be adding a couple more features.

1. Add the ability to add your own subnets in the Auto-Assign from Range section
![img-2023-07-18-13-06-26](https://github.com/sinamics/next_ztnet/assets/56016218/ced785ce-06f3-4ede-92cd-156f7b7637a1)

2. It can be very useful to know the ip of the peer from which it connects. (maby in conn status column)
![img-2023-07-18-13-21-09](https://github.com/sinamics/next_ztnet/assets/56016218/45a7a76b-2f11-446d-9f93-0f6bde184023)

**Comment by sinamics** [comment]
Great. 
Yes, these are usefull as well.
I will put them on the todo list.

**Comment by sinamics** [comment]
@dsm150 just pushed version 0.1.4 with physical ip and ability to add your own ip range.
it also have some general UX improvements.

Just keep in mind that if the link is Relayed, the external IP is not available.

**Comment by dsm150** [comment]
@sinamics all working flawlessly!!!!
Now I will try to add more networks and peers.
You have created a great solution!

**Comment by dsm150** [comment]
Updated to version 0.2.0 - everything went smoothly, it works properly.

**Comment by sinamics** [comment]
@dsm150 thx for the feedback, glad to hear it works fine.

**Comment by dsm150** [comment]
one more question
how i can manage "Allow Management From:" in settings?

**Comment by sinamics** [comment]
that value is written in the docker-composer file.
Its the network all containers are connected to, hence the cidr 29 that allows 6 IP`s 

`ZT_ALLOW_MANAGEMENT_FROM=172.31.255.0/29`

```yml
  zerotier:
    image: zyclonite/zerotier:1.10.6
    hostname: zerotier
    container_name: zerotier
    restart: unless-stopped
    volumes:
      - zerotier:/var/lib/zerotier-one
    cap_add:
      - NET_ADMIN
      - SYS_ADMIN
    devices:
      - /dev/net/tun:/dev/net/tun
    networks:
      - app-network
    ports:
      - "9993:9993/udp"
    environment:
      - ZT_OVERRIDE_LOCAL_CONF=true
      - ZT_ALLOW_MANAGEMENT_FROM=172.31.255.0/29
```

**Comment by sinamics** [comment]
For information, pushed v0.2.1, it has some improvements if you have many networks.
tested with 10 000 networks and it works just fine now.

https://github.com/sinamics/next_ztnet/releases/tag/v0.2.1

**Comment by sinamics** [comment]
I would also appreciate your suggestions about other features you might be interested in having added to the application.

I had these on my mind:
- ZeroNSD appears to be an exciting feature. However, its implementation could be challenging, as it requires launching a service for each network, which needs careful handling.
- Member Tags to ease the filtration if you have many members.

**Comment by dsm150** [comment]
ZeroNSD would probably be useful, but personally I have not yet experienced the need to use this technology.
If the implementation is possible, and does not require a titanic effort, of course the community will gladly accept it.

Member Tags - definitely useful!

Still might consider adding API functions

I also saw controller implementations along with the Planet server. This is an interesting thing to create your own network, independent of the developer's servers.

By the way, here is a link to my post about your development)
[My post](https://discuss.zerotier.com/t/zerotier-controller-ui-next-ztnet/14854)

**Comment by sinamics** [comment]
> ZeroNSD would probably be useful, but personally I have not yet experienced the need to use this technology.
If the implementation is possible, and does not require a titanic effort, of course the community will gladly accept it.

This is achievable but requires a thoughtful and intuitive approach. It's not as straightforward as it may appear. I'll examine this further when time permits.

> Member Tags - definitely useful!

Almost done. 
I've incorporated the "Annotation" feature into the member options menu. I believe it's a neat addition.

> I also saw controller implementations along with the Planet server. This is an interesting thing to create your own network, independent of the developer's servers.

Indeed, it's an aspect worth exploring further.

> By the way, here is a link to my post about your development)

Thx. Appreciated.

**Comment by sinamics** [comment]
pushed version 0.2.4.
!NOTE: I also renamed repo to "ztnet"
https://github.com/sinamics/ztnet/releases/tag/v0.2.4

would you mind update the link in zerotier forum?
https://github.com/sinamics/ztnet

---
New feature where you can add notation to a member. 
I also added option in admin settings to show notation tag or enable table row background color in table. 

I did not add a Filter option in table view, as you can just search the Notation you want.

**Comment by dsm150** [comment]
Sometimes, for security reasons, we do this:
We connect the controller to our own, and generally close everything from the outside, except for port 9993. This is especially true when the controller is located in a data center and has a static white IP.
As a result, they should be able to manage the controller and the operating system only by connecting to the network.
I tried to do this, but something went wrong..
The controller has connected, the command line shows that everything is OK, but the web interface shows that the peer is offline. also, I can't ping from controller other members of the network.

![img-2023-07-30-18-29-48](https://github.com/sinamics/ztnet/assets/56016218/80ce1c3b-0009-4d6a-8cee-dc28924216c2)
![img-2023-07-30-18-29-23](https://github.com/sinamics/ztnet/assets/56016218/e8ac355e-7e1e-429c-b898-910499c928b1)
![img-2023-07-30-18-35-50](https://github.com/sinamics/ztnet/assets/56016218/369575d8-f961-4b8b-aa15-5911ff335a18)

**Comment by sinamics** [comment]
@dsm150 not sure i understood your example, but ztnet looks for the controller at `http://zerotier:9993`, where `zerotier` is the container name of the controller.
you might want to change this value in docker-compose to point to your custom zt controller.

```yml
environment:
   ZT_ADDR: http://zerotier:9993
```

Or i have might misunderstanded you.

**Comment by dec0dOS** [comment]
Hello, @sinamics!

[zero-ui](http://github.com/dec0dOS/zero-ui/) maintainer is here. 
I glad to see the new ZeroTier controller UI option!
I'm thrilled with the selected technologies and the work itself, It looks awesome!

By the way, I couldn't find any information about the ZeroTier Central API compatibility. Are you planning to support it?

**Comment by dsm150** [comment]
@sinamics  Let me try again to explain what I mean.
In the picture, this is just the implementation of the controller of the respected @dec0dOS, it can be seen that a regular client and the controller itself are connected to their own network. This allows you to limit access to the controller itself only from the ZeroTier network.
![img-2023-07-31-00-53-47](https://github.com/sinamics/ztnet/assets/56016218/6fe83f87-0836-4282-a537-6ebf29c045be)

**Comment by dec0dOS** [comment]
Something went wrong with the latest release. I've used the docker-compose.yml provided in the README.
![Screenshot 2023-07-30 at 23 12 47](https://github.com/sinamics/ztnet/assets/14824142/b7419594-4b78-44f8-be25-62ff7220fd6a)

**Comment by dec0dOS** [comment]
![Screenshot 2023-07-30 at 23 17 40](https://github.com/sinamics/ztnet/assets/14824142/d7a956f4-fdcd-4f21-be63-757d27b83aea)

IPv4 Assignment card does not scale well

**Comment by dec0dOS** [comment]
Modals with Dark Reader enabled are transparent, seems to be it is better to check the models for compatibility. (also noticed some other issues with models in Safari)

![Screenshot 2023-07-30 at 23 23 30](https://github.com/sinamics/ztnet/assets/14824142/61cae142-5e2b-4c94-be9a-4dc6c2af90a2)

**Comment by dec0dOS** [comment]
Any news about the IPv6 support? That is a highly anticipated feature for ZeroUI.

I really enjoyed using ZTNET. That is exactly what I would like ZeroUI to be like, but didn't have enough bandwidth at that time to fully implement all the features. I'm still a heavy ZeroTier user, so it's great to see the community continue developing other solutions.
Thanks again for your awesome work, @sinamics!

**Comment by sinamics** [comment]
@dsm150 Ah, yes now i understand. This seems like a very neat solution. I will test this myself and get back to you.

**Comment by sinamics** [comment]
@dec0dOS Hi. Glad you liked it, you have done a great job yourself with ZeroUI.  

> By the way, I couldn't find any information about the ZeroTier Central API compatibility. Are you planning to support it?  

It's interesting that you've brought this up. I actually started researching on central API compatibility a few days ago, so yes, I'm working on adding it.

Thanks for all the testing you've done. I'll fix the errors and scaling issues you found for the next update. Much appreciated.
As for the IPv6 support, I'm interested in adding it too, but I'll need to study up a bit on IPv6 first because I'm not too familiar with it.

Overall, I'm trying to add as many useful features as I can to make ztnet more flexible.

**Comment by sinamics** [comment]
FYI
just pushed version 0.3.0 with Zerotier Central API support.
https://github.com/sinamics/ztnet/releases/tag/v0.3.0

**Comment by dec0dOS** [comment]
Hey @sinamics! That is awesome!
Seems to be you could use the ZTNET to manage the ZeroTier Central. But is it possible to manage the ZTCUI networks with ZeroTier Central-like API?

The great ZeroTier (or any API-based mesh networks) feature is that you could set up your network programmatically. If some want to migrate from ZeroTier Central or from ZeroUI to ZTNET, for some advanced users it requires to rewrite the API interaction.

**Comment by sinamics** [comment]
ztnet does not have a any public api endpoint, but will probably add it further down the road. 
I agree that this will make it easier for users to switch platforms.

**Comment by jary888** [comment]
@sinamics     
Hi ,   v0.31  NEW: When migrating from a custom ZeroTier controller, there will be a new table in admin settings to manually assign existing networks to an admin member.    How can this be achieved?😢😢😢

**Comment by sinamics** [comment]
@jary888  Could you please clarify your objectives?

Should you wish to utilize a controller different from `zyclonite/zerotier`, it's essential to modify the ZT_ADDR & ZT_SECRET values in docker-compose to point to your desired controller:

```yml
    environment:
      ZT_ADDR: http://zt_controller_ip:9993
      ZT_SECRET: (content of authtoken.secret)
```

Additionally, if the controller you've set up has multiple networks, ensure you map these to a user within ztnet. Once mapped, they'll appear under the Local Network Page.
All unlinked networks will be in shown in a table (Admin page => Controller)
![image](https://github.com/sinamics/ztnet/assets/7771916/8d33bb9d-13b9-4073-9008-1011bcca024d)


The reason you need to map networks, is that ztnet has multiuser support and all networks with its members needs to be linked to a registerd user.

If you still have problem, please create a new issue.

**Comment by dsm150** [comment]
I suggest adding the "last seen" function to the "Conn status" field.
This would be a very useful feature

**Comment by dec0dOS** [comment]
> The great ZeroTier (or any API-based mesh networks) feature is that you could set up your network programmatically. If some want to migrate from ZeroTier Central or from ZeroUI to ZTNET, for some advanced users it requires to rewrite the API interaction.

By the way, the compatible API could be used in such cases as terraform integration, where you could describe your network in a common way
https://www.zerotier.com/blog/zerotier-central-now-integrates-with-hashicorp-terraform/

Thanks you for your work, @sinamics! Ztnet becomes better every week!

**Comment by sinamics** [comment]
Interresting. I'm using terraform myself for other projects, and it's really cool to deploy as code. I'll definitely look into that when time permits.

**Comment by dsm150** [comment]
> @sinamics Let me try again to explain what I mean. In the picture, this is just the implementation of the controller of the respected @dec0dOS, it can be seen that a regular client and the controller itself are connected to their own network. This allows you to limit access to the controller itself only from the ZeroTier network. ![img-2023-07-31-00-53-47](https://user-images.githubusercontent.com/56016218/257078027-6fe83f87-0836-4282-a537-6ebf29c045be.png)

@sinamics We are looking forward to this feature, I will be grateful if you find time for this. 🙏🏻
